spec/models/question_mover_spec.rb

```diff
@@ -0,0 +1,316 @@
+      expect { mover.move }.not_to raise_error
+      question.reload
+      target.reload
+      expect(question.parent.data_type).to eq("group")
+
+      case relation
+      when :before
+        expect(question.parent).to eq(target.parent)
+        expect(question.position).to eq(target.position - 1)
+      when :after
+        expect(question.parent).to eq(target.parent)
+        expect(question.position).to eq(target.position + 1)
+      when :inside
+        expect(question.parent).to eq(target)
+        expect(question.position).to eq(0)
+      else
+        raise "Invalid relation"
+      end
+    end
+  end
+
+  shared_examples_for "denies" do
+    it "denies with exception" do
+      expect { mover.move }.to raise_error(ArgumentError)
+    end
+  end
+
+  context "moving question with current division" do
+    context "within same parent" do
+      context "to top of division block" do
+        let(:current_division) { div_c1 }
+        let(:question) { q114 }
+        let(:target) { q113 }
+        let(:relation) { :before }
+        it_behaves_like "succeeds"
+      end
+
+      context "to bottom of division block" do
+        let(:current_division) { div_c1 }
+        let(:question) { q113 }
+        let(:target) { q115 }
+        let(:relation) { :after }
+        it_behaves_like "succeeds"
+      end
+
+      context "outside of division block" do
+        let(:current_division) { div_c1 }
+        let(:question) { q113 }
+        let(:target) { q116 }
+        let(:relation) { :after }
+        let(:error) { "must be adjacent to questions of same division" }
+        it_behaves_like "denies"
+      end
+    end
+
+    context "to different parent" do
+      context "to non-group" do
+        let(:current_division) { div_c1 }
+        let(:question) { q113 }
+        let(:target) { q114 }
+        let(:relation) { :inside }
+        let(:error) { "parent must be group" }
+        it_behaves_like "denies"
+      end
+
+      context "to parent with lower division" do
+        let(:current_division) { div_a }
+        let(:question) { q2 }
+        let(:target) { q4 }
+        let(:relation) { :inside }
+        let(:error) { "must have parent of same or ancestor division" }
+        it_behaves_like "denies"
+      end
+
+      context "to parent with ancestor division" do
+        context "when existing division block exists" do
+          context "within existing division block :after target with same division" do
+            let(:current_division) { div_c1 }
+            let(:question) { q41 }
+            let(:target) { q113 }
+            let(:relation) { :after }
+            it_behaves_like "succeeds"
+          end
+
+          context "within existing division block :after target with different division" do
+            let(:current_division) { div_c1 }
+            let(:question) { q41 }
+            let(:target) { q112 } # has div_c3
+            let(:relation) { :after }
+            it_behaves_like "succeeds"
```

Comment:
Am I following right that it's ok to have a q for div B be after qs for div C as long as Div B and C are siblings? (re error message: "must be adjacent to questions of same division depth")

Reply:
Yes exactly. We just enforce that each division is a contiguous block in that case, so the numbering doesn't get _too_ weird.